### ISSUE TITLE
SymbolModality wrongly dropping out entire input?

### ISSUE THREAD

**Issue by cbockman** [tensorflow/tensor2tensor#590]
I posted this on gitter, but no reply, so wanted to surface this:

A recent update added dropout to SymbolModality....on the input itself: https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/modalities.py#L101
Was this intentional? It seems kind of strange.

Dropout after embedding I understand but on the input itself is something that is pretty rare in the literature (at least in the text space...maybe SymbolModality is being used internally to support cases where this does make a lot of sense?).

That said, if I'm using SymbolModality to embed integer tokens (e.g., https://github.com/tensorflow/tensor2tensor/blob/a1d7ed7e7d96ea0d89ca153ba46b7f51d394580f/docs/overview.md), seems odd to drop them out entirely (at least as the baseline, default option--I realize that there are perhaps research reasons why one might want to try this).

For context, my mental model here is using SymbolModality to support ingest tokenized (to integers) words or subwords.

**Comment by martinpopel** [comment]
Just for the record, @lukaszkaiser answered you on Gitter "*symbol dropout is a good idea, try it out!*"

**Comment by cbockman** [comment]
Thanks @martinpopel !  I see that, but I don't think it gets quite at my concern?

This makes it so, e.g., in any sort of language problem, we (I think) just drop out entire words.  This is 1) a very large change in baseline behavior and 2) non-standard in the deep learning literature.

Is this intended?  The change is sufficiently radical that I wonder if it was intended for dropout to simply be added on the post-embedding.

It could be that SymbolModality isn't really intended for the use case I've described?

Or it could be I'm dramatically misinterpreting what is going on in the code.  :)

**Comment by colmantse** [comment]
Dropping out on symbol make sense on word, subword and char level. For char and subword level, the model is more robust towards miss spelling and stuff. For word level, it learns to infer meaning even if some words provided not keywords are dropped.




取得 Outlook for Android









On Fri, Feb 16, 2018 at 5:16 AM +0800, "cbockman" <notifications@github.com> wrote:












Thanks @martinpopel !  I see that, but I don't think it gets quite at my concern?


This makes it so, e.g., in any sort of language problem, we (I think) just drop out entire words.  This is 1) a very large change in baseline behavior and 2) non-standard in the deep learning literature.


Is this intended?  The change is sufficiently radical that I wonder if it was intended for dropout to simply be added on the post-embedding.


It could be that SymbolModality isn't really intended for the use case I've described?


Or it could be I'm dramatically misinterpreting what is going on in the code.  :)

**Comment by cbockman** [comment]
I have yet to see any SOTA models in the literature do this for any of the most common NLP subclasses of problems (translation, natural language inference, reading comprehension/QA, etc.).  

Conceptually, this is also counter-intuitive--dropping out a key word and forcing a translation to occur based on that word being gone is conceptually very different than dropping out neural units to prevent co-adaptation.

(And SymbolModality does appear to be at the base of the translation problem implementation: https://github.com/tensorflow/tensor2tensor/blob/0d464ff699029cd604fdb11fb532f64bebc9134c/tensor2tensor/data_generators/problem.py#L1063)

If the t2t team states that this is 100% intentional, then, great, I'll just work around this.  :)  And/or perhaps this really does work great in practice!

But, again, I can't think of any literature offhand where this is either the baseline behavior, or where this behavior was shown to generally improve outcomes in the broad array of NLP activities.  I can certainly understand wanting to research behavior like this, but it--currently--seems implausible that this is really the best default behavior, to the degree that I'd love affirmation that this really is intended, versus the dropout being intended to be *post* embedding (which is common).

**Comment by colmantse** [comment]
Oh sorry, I am not from the t2t team, was just an opinion, should have experiment with it first. I agree dropping keyword would be a problem. I just think it benefit subword and char model more.




取得 Outlook for Android









On Fri, Feb 16, 2018 at 3:53 PM +0800, "cbockman" <notifications@github.com> wrote:












I have yet to see any SOTA models in the literature do this for any of the most common NLP subclasses of problems (translation, natural language inference, reading comprehension/QA, etc.).


Conceptually, this is also counter-intuitive--dropping out a key word and forcing a translation to occur based on that word being gone is conceptually very different than dropping out neural units to prevent co-adaptation.


(And SymbolModality does appear to be at the base of the translation problem implementation: https://github.com/tensorflow/tensor2tensor/blob/0d464ff699029cd604fdb11fb532f64bebc9134c/tensor2tensor/data_generators/problem.py#L1063)


If the t2t team states that this is 100% intentional, then, great, I'll just work around this.  :)  And/or perhaps this really does work great in practice!


But, again, I can't think of any literature offhand where this is either the baseline behavior, or where this behavior was shown to generally improve outcomes in the broad array of NLP activities.  I can certainly understand wanting to research behavior like this, but it--currently--seems implausible that this is really the best default behavior, to the degree that I'd love affirmation that this really is intended, versus the dropout being intended to be post embedding (which is common).

**Comment by cbockman** [comment]
Perhaps because subwords are the default way everything is built, this is
notionally intentional  But I would argue that this dropout should be
occurring in one of the Text-level Problems, not at the SymbolModality, as
it should be kept generic to support--as its name would suggest--all
reasonable possible downstream symbol modality uses (to include, e.g., word
embeddings).

**Comment by colmantse** [comment]
That I agree, it would better be a selectable feature rather than default.

Sent from Mail for Windows 10

**Comment by raphael-sch** [comment]
It is useful to train a model as denoising autoencoder
